src/kdalgorithms.h

```diff
@@ -33,18 +34,25 @@
 
 namespace kdalgorithms {
 
+struct invalid_self_copy : public std::exception
+{
+    const char *what() const throw()
+    {
+        return "Self copy on container without reserve method not supported";
+    }
+};
+
 // -------------------- copy --------------------
 template <typename InputContainer, typename OutputContainer>
 #if __cplusplus >= 202002L
 requires ContainerOfType<OutputContainer, ValueType<InputContainer>>
 #endif
 void copy(InputContainer &&input, OutputContainer &output)
 {
-#if __cplusplus >= 201703L
-    if constexpr (detail::has_reserve_method_v<OutputContainer>) {
-        output.reserve(input.size() + output.size());
+    bool foundReserve = detail::reserve(output, input.size() + output.size());
+    if (!foundReserve && detail::is_at_same_address(input, output)) {
+        throw invalid_self_copy();
```

Comment:
Would it make any sense, as alternative strategy, to create another container; copy input+output in there; and move-assign this new container onto input once done?